### ISSUE TITLE
Flexible Cache / Caching Policy

### ISSUE THREAD

**Issue by snodnipper** [square/okhttp#1605]
It would be great if devs had a cache interface so they could define a custom cache implementation. @swankjesse suggested this requirement may be better served (or certainly complemented) by a caching policy and asked for a separate issue to be raised.

Some use cases where a custom cache would be useful:
- offline mode: user selects certain content (URLs) to be available offline - which overrides any default cache eviction behaviour.  This could be useful for media content etc. _think Spotify’s offline mode for specific content_.
  - I have had some success by abstracting two caches (offline/steaming) with deferent properties.  The offline bucket could only be evicted by the user whereas streaming behaved normally.
- abstracting multiple storage locations: I have been working on a project where users have cached content directories of several GiB, sometimes on different volumes (which can be unmounted etc.).  I have written a cache wrapper to ensure the desired behaviour.
- sensitive data: some content requires encryption, others do not.

My immediate thoughts are that a contract similar to the following could be useful.

```
DataProvider
void PUT(String url, byte[] bytes);
byte[] GET(String url);
```

Without too much thought I can imagine a chain of responsibility pattern for caching: MemoryDataProvider -> DiskDataProvider -> NetworkDataProvider

Related:
https://github.com/square/okhttp/issues/1021
https://github.com/square/okhttp/issues/1459

**Comment by swankjesse** [comment]
If you'd like different storage locations, you can already do that by using two OkHttpClient instances, each configured with a different cache.

For offline mode, it's a little bit weird to use the HTTP cache to power that. In particular, because you need to be extremely careful around how you seed the cache, and what URLs your application attempts to fetch. Doing this in the application layer might be a better fit.

**Comment by snodnipper** [comment]
agreed.  Different OkHttpClient instances certainly could get the job done - and potentially, the offline instance could simply have an extremely large cache size with headers to not expire anytime soon / accept anything in the bucket.  Cheers.

**Comment by mandrachek** [comment]
The sensitive data bit is what I need. It would be so lovely to be able to transparently encrypt/decrypt cached responses on the fly (HIPAA stuff - has to be encrypted on the wire and "at rest").  

So it would be good if we could intercept the read and writes of cache entries (and keys!). At this point I'm stuck manually caching stuff. :(

**Comment by swankjesse** [comment]
If you're feeling adventurous and want to depend on an unstable internal API, take a look at these two APIs:
https://github.com/square/okhttp/blob/master/okhttp/src/main/java/com/squareup/okhttp/internal/DiskLruCache.java#L245
https://github.com/square/okhttp/blob/master/okhttp/src/main/java/com/squareup/okhttp/Cache.java#L171

No promises, but it might unblock you.

**Comment by wuwu2000** [comment]
Just for the next user reading this. Because package changed, URLs are broken now
https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/internal/DiskLruCache.java
https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/Cache.java

**Comment by JakeWharton** [comment]
Careful because your links will break when we update to version 4! Always
press 'y' before copying a GitHub link to a source file.

**Comment by sberg413** [comment]
@swankjesse  - as you alluded to in the wishlist thread, encrypting cached data can be accomplished by implementing a custom FileSystem that wraps the IO streams in cipher streams. I was able to get a POC of this working using the 3-arg constructor of Cache via reflection. Why not make this constructor public? Is there a major downside or concern in allowing for a custom FIleSystem implementation?  Thanks.

**Comment by swankjesse** [comment]
It dramatically increases the amount of public API we’re exporting.

**Comment by yschimke** [comment]
@swankjesse Could a middle ground be to provide a factor method inside internal to call that API without reflection tricks.  It would still be brittle non-public API with clear warnings against using (this is not even considered experimental etc), but would let us have a cleaner option for people who are already abusing the internal API of FileSystem.

I think there are cases where some users would manage okhttp3 versions in their app carefully and would accept the upgrade pain if we massively refactor.

**Comment by swankjesse** [comment]
Biggest challenge is what that API looks like.

With our filesystem cache the headers and body are a single file and we assume cheap random access, atomicity, and reliability. Beyond a single entry we assume very fast presence checks which are separate from eligible checks.

With our planned relay cache we further assume reading partial writes. We do awkward things like serialize certain operations on the cache.

If we extract an interface we either commit to these design assumptions or introduce more flexibility.

I’d like to do it all! But it's a big undertaking. Especially if there are other concurrent users of the cache.

**Comment by yschimke** [comment]
Maybe one to slow boil, introduce a draft PR and get feedback on possible designs compared to the current design.  Introduce as part of 4.0?

**Comment by swankjesse** [comment]
Yes! One other option is to provide a cache construction kit: public APIs to decide what's eligible to be saved and loaded, and let users define interceptors that do the real work.

**Comment by swankjesse** [comment]
Ie. “library” not “framework” caching.

**Comment by kenyee** [comment]
Our unusual use case is we want the cache to strip a queryparam out when storing/looking at something.  Use used to do this by implementing the Cache interface but it looks to be final now :-(
Any suggestions?

**Comment by tsy12321** [comment]
> Our unusual use case is we want the cache to strip a queryparam out when storing/looking at something. Use used to do this by implementing the Cache interface but it looks to be final now :-(
> Any suggestions?

I hava the same question.... @swankjesse

**Comment by yschimke** [comment]
Closing for now, no plans. If we did this it should probably be a clean simple API and allow external implementations.
